File: CS/SimpleWebReportCatalog/ReportStorage.cs
using System;
using System.Data;
using SimpleWebReportCatalog.CatalogDataTableAdapters;
// ...

namespace SimpleWebReportCatalog {
    public class ReportStorage {
        private CatalogData catalogDataSet;
        private DataTable reportsTable;
        private ReportsTableAdapter reportsTableAdapter;

        public ReportStorage() {
            catalogDataSet = new CatalogData();
            reportsTableAdapter = new ReportsTableAdapter();
            reportsTableAdapter.Fill(catalogDataSet.Reports);

            reportsTable = catalogDataSet.Tables["Reports"];
        }

        public void WriteReport(ReportModel reportModel) {
            DataRow row = reportsTable.NewRow();

            row["ReportID"] = reportModel.ReportId;
            row["LayoutData"] = reportModel.LayoutData;

            reportsTable.Rows.Add(row);
            reportsTableAdapter.Update(catalogDataSet);
            catalogDataSet.AcceptChanges();
        }


        public ReportModel GetReport(string reportId) {
            DataRow row = reportsTable.Rows.Find(reportId);

            if(row != null) {
                ReportModel model = new ReportModel();
                model.ReportId = (string)row["ReportID"];
                model.LayoutData = (Byte[])row["LayoutData"];
                return model;
            } else {
                return null;
            }
        }

        public void UpdateReport(ReportModel reportModel) {
            DataRow row = reportsTable.Rows.Find(reportModel.ReportId);

            if(row != null) {
                row["LayoutData"] = reportModel.LayoutData;
                reportsTableAdapter.Update(catalogDataSet);
                catalogDataSet.AcceptChanges();
            }
        }

        public void RemoveReport(string reportId) {
            DataRow row = reportsTable.Rows.Find(reportId);

            if(row != null) {
                row.Delete();
                reportsTableAdapter.Update(catalogDataSet);
                catalogDataSet.AcceptChanges();
            }
        }
    }
}

File: VB/SimpleWebReportCatalog/ReportStorage.vb
Imports System
Imports System.Data
Imports SimpleWebReportCatalog.CatalogDataTableAdapters
Imports CatalogDataTableAdapters

' ...

Namespace SimpleWebReportCatalog
    Public Class ReportStorage
        Private catalogDataSet As CatalogData
        Private reportsTable As DataTable
        Private reportsTableAdapter As ReportsTableAdapter


        Public Sub New()
            catalogDataSet = New CatalogData()
            reportsTableAdapter = New ReportsTableAdapter()
            reportsTableAdapter.Fill(catalogDataSet.Reports)

            reportsTable = catalogDataSet.Tables("Reports")
        End Sub


        Public Sub WriteReport(ByVal reportModel As ReportModel)
            Dim row As DataRow = reportsTable.NewRow()

            row("ReportID") = reportModel.ReportId
            row("LayoutData") = reportModel.LayoutData

            reportsTable.Rows.Add(row)
            reportsTableAdapter.Update(catalogDataSet)
            catalogDataSet.AcceptChanges()
        End Sub


        Public Function GetReport(ByVal reportId As String) As ReportModel
            Dim row As DataRow = reportsTable.Rows.Find(reportId)

            If row IsNot Nothing Then
                Dim model As New ReportModel()
                model.ReportId = DirectCast(row("ReportID"), String)
                model.LayoutData = DirectCast(row("LayoutData"), Byte())
                Return model
            Else
                Return Nothing
            End If
        End Function


        Public Sub UpdateReport(ByVal reportModel As ReportModel)
            Dim row As DataRow = reportsTable.Rows.Find(reportModel.ReportId)

            If row IsNot Nothing Then
                row("LayoutData") = reportModel.LayoutData
                reportsTableAdapter.Update(catalogDataSet)
                catalogDataSet.AcceptChanges()
            End If
        End Sub


        Public Sub RemoveReport(ByVal reportId As String)
            Dim row As DataRow = reportsTable.Rows.Find(reportId)

            If row IsNot Nothing Then
                row.Delete()
                reportsTableAdapter.Update(catalogDataSet)
                catalogDataSet.AcceptChanges()
            End If
        End Sub
    End Class
End Namespace







